pkg/vere/main.c

```diff
@@ -1726,6 +1727,165 @@ _cw_prep(c3_i argc, c3_c* argv[])
+    exit(1);
+  }
+
+  // gracefully shutdown the pier if it's running
+  u3_disk* old_u = _cw_disk_init(u3_Host.dir_c);
+
+  // note: this include patch applications (if any)
+  u3m_boot(u3_Host.dir_c, (size_t)1 << u3_Host.ops_u.lom_y);
+
+  // check if there's a *current* snapshot
+  if ( old_u->dun_d != u3A->eve_d ) {
+    fprintf(stderr, "chop: error: snapshot is out of date, please "
+                    "start/shutdown your pier gracefully first\r\n");
+    fprintf(stderr, "chop: eve_d: %" PRIu64 ", dun_d: %" PRIu64 "\r\n", u3A->eve_d, old_u->dun_d);
+    exit(1);
+  }
+
+  u3e_backup(c3y);  //  backup current snapshot
+
+  // initialize the lmdb environment
+  // see disk.c:885
+  const size_t siz_i =
+  // 500 GiB is as large as musl on aarch64 wants to allow
+  #if (defined(U3_CPU_aarch64) && defined(U3_OS_linux))
+    0x7d00000000;
+  #else
+    0x10000000000;
+  #endif
+  c3_c log_c[8193];
+  snprintf(log_c, sizeof(log_c), "%s/.urb/log", u3_Host.dir_c);
+
+  // get the first/last event numbers from the event log
+  c3_d fir_d, las_d;
+  if ( c3n == u3_lmdb_gulf(old_u->mdb_u, &fir_d, &las_d) ) {
+    fprintf(stderr, "chop: failed to load latest event from database\r\n");
+    exit(1);
+  }
+
+  // get the metadata
+  c3_d     who_d[2];
+  c3_o     fak_o;
+  c3_w     lif_w;
+  if ( c3y != u3_disk_read_meta(old_u->mdb_u, who_d, &fak_o, &lif_w) ) {
+    fprintf(stderr, "chop: failed to read metadata\r\n");
+    exit(1);
+  }
+
+  // get the last event
+  u3_lmdb_walk itr_u;
+  size_t       len_i;
+  void*        buf_v[1];
+  if ( c3n == u3_lmdb_walk_init(old_u->mdb_u, &itr_u, las_d, las_d) ) {
+    fprintf(stderr, "chop: failed to initialize iterator\r\n");
+    exit(1);
+  }
+  if ( c3n == u3_lmdb_walk_next(&itr_u, &len_i, buf_v) ) {
+    fprintf(stderr, "chop: failed to read event\r\n");
+    exit(1);
+  }
+  u3_lmdb_walk_done(&itr_u);
+
+  // initialize a fresh lmdb environment in the "chop" subdir
+  c3_c cho_c[8193];
+  snprintf(cho_c, sizeof(cho_c), "%s/chop", log_c);
+  c3_mkdir(cho_c, 0700);
+  MDB_env* new_u = u3_lmdb_init(cho_c, siz_i);
+
+  // write the metadata to the database
+  if ( c3n == u3_disk_save_meta(new_u, who_d, fak_o, lif_w) ) {
+    fprintf(stderr, "chop: failed to save metadata\r\n");
+    exit(1);
+  }
+
+  // write the last event to the database
+  // warning: this relies on the old database still being open
+  if ( c3n == u3_lmdb_save(new_u, las_d, 1, buf_v, &len_i) ) {
+    fprintf(stderr, "king: failed to write last event\r\n");
+    exit(1);
+  }
+
+  // backup the original database file
+  c3_c dat_c[8193], bak_c[8193];
+  snprintf(dat_c, sizeof(dat_c), "%s/data.mdb", log_c);
+  // "data_<first>-<last>.mdb.bak"
+  snprintf(bak_c, sizeof(bak_c), "%s/data_%" PRIu64 "-%" PRIu64 ".mdb.bak", cho_c, fir_d, las_d);
+  c3_rename(dat_c, bak_c);
+
+  // rename new database file to be official
+  c3_c new_c[8193];
+  snprintf(new_c, sizeof(new_c), "%s/data.mdb", cho_c);
+  c3_rename(new_c, dat_c);
```

Comment:
Both of these `rename()` calls need error handling/printfs.

Reply:
You can't just compare against `errno` here, you have to check the return code from `rename()`.